Add WKViewCreateWPE for PlayStation

donny-dont opened the pull request:
#### 0054d6405e19aafc6d5592bffd6f9aa707b008b0
<pre>
Add WKViewCreateWPE for PlayStation
<a href="https://bugs.webkit.org/show_bug.cgi?id=257660">https://bugs.webkit.org/show_bug.cgi?id=257660</a>

Reviewed by Fujii Hironori.

Add `WKViewCreateWPE` so a `wpe_view_backend` can be passed into the
underlying `PlayStationWebView`. This is only relevant when
`USE(WPE_BACKEND_PLAYSTATION)` is `ON`.

* Source/WebKit/UIProcess/API/C/playstation/WKView.cpp:
(WKViewCreate):
(WKViewCreateWPE):
* Source/WebKit/UIProcess/API/C/playstation/WKView.h:
* Source/WebKit/UIProcess/playstation/PageClientImpl.cpp:
(WebKit::PageClientImpl::hostFileDescriptor):
* Source/WebKit/UIProcess/playstation/PlayStationWebView.cpp:
(WebKit::PlayStationWebView::create):
(WebKit::PlayStationWebView::PlayStationWebView):
(WebKit::m_backend):
* Source/WebKit/UIProcess/playstation/PlayStationWebView.h:
(WebKit::PlayStationWebView::backend):
* Tools/MiniBrowser/playstation/CMakeLists.txt:
* Tools/MiniBrowser/playstation/WebViewWindow.cpp:
(WebViewWindow::WebViewWindow):
* Tools/MiniBrowser/playstation/WebViewWindow.h:

Canonical link: <a href="https://commits.webkit.org/264877@main">https://commits.webkit.org/264877@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0aa4c943827f78fb8d6fba46b74143236000068

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8997 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9285 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9503 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10650 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/8963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9005 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/11272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/9253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/11809 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9143 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/11272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10809 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/11272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/8229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/15706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/11272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/8377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/11693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/8869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/9253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/8119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/12334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1038 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/8619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebKit/UIProcess/API/C/playstation/WKView.cpp:
```diff
@@ -56,7 +56,20 @@ WKCursorType toWKCursorType(const WebCore::Cursor& cursor)
 WKViewRef WKViewCreate(WKPageConfigurationRef configuration)
 {
+#if USE(WPE_BACKEND_PLAYSTATION)
+    RELEASE_ASSERT_WITH_MESSAGE(false, "API unavailable with WPE Backend PlayStation");
```
It's nonsense to expose this API if this API just crashes.
Why don't you make this API fail?

fujii replied:
Or, not exposing this API if `USE(WPE_BACKEND_PLAYSTATION)`.

donny-dont replied:
I didn't see any `ENABLE` or `USE` in the WK API headers since those are internal definitions. That's why I made the API that.

In this one if there's no backend it'll crash later on so better to crash with a warning.

The long term goal is for this API to be deprecated and then replaced with the WPE one.

fujii replied:
Got it.

fujii commented on Source/WebKit/UIProcess/API/C/playstation/WKView.cpp:
```diff
@@ -56,7 +56,20 @@ WKCursorType toWKCursorType(const WebCore::Cursor& cursor)
+WKViewRef WKViewCreateWPE(struct wpe_view_backend* backend, WKPageConfigurationRef configuration)
+{
+#if USE(WPE_BACKEND_PLAYSTATION)
+    return WebKit::toAPI(WebKit::PlayStationWebView::create(backend, *WebKit::toImpl(configuration)).leakRef());
```
I think you shoudn't leakRef here. Why do you leak here?

fujii reviewed (commented):
Why does the style checker complain?
> ERROR: Source/WebKit/UIProcess/playstation/PlayStationWebView.cpp:49:  Code inside a namespace should not be indented.  [whitespace/indent] [4]

donny-dont commented on Source/WebKit/UIProcess/playstation/PlayStationWebView.cpp:
```diff
@@ -30,8 +30,34 @@
+
+PlayStationWebView::PlayStationWebView(struct wpe_view_backend* backend, const API::PageConfiguration& conf)
+    : m_pageClient(makeUnique<PageClientImpl>(*this))
+    , m_viewStateFlags { WebCore::ActivityState::WindowIsActive, WebCore::ActivityState::IsFocused, WebCore::ActivityState::IsVisible, WebCore::ActivityState::IsInWindow }
```
@fujii I believe this is the issue the style checker is taking issue with.

donny-dont commented on Source/WebKit/UIProcess/API/C/playstation/WKView.cpp:
```diff
@@ -56,7 +56,20 @@ WKCursorType toWKCursorType(const WebCore::Cursor& cursor)
+#if USE(WPE_BACKEND_PLAYSTATION)
+    RELEASE_ASSERT_WITH_MESSAGE(false, "API unavailable with WPE Backend PlayStation");
+#else
     return WebKit::toAPI(WebKit::PlayStationWebView::create(*WebKit::toImpl(configuration)).leakRef());
```
@fujii the `leakRef` is coming from here.

fujii replied:
WK API convention is that not leaking here.

donny-dont replied:
`WebKit::PlayStationWebView::create` is returning a `RefPtr` so it appears this is the way.

donny-dont committed:
Add WKViewCreateWPE for PlayStation
https://bugs.webkit.org/show_bug.cgi?id=257660

Reviewed by Fujii Hironori.

Add `WKViewCreateWPE` so a `wpe_view_backend` can be passed into the
underlying `PlayStationWebView`. This is only relevant when
`USE(WPE_BACKEND_PLAYSTATION)` is `ON`.

* Source/WebKit/UIProcess/API/C/playstation/WKView.cpp:
(WKViewCreate):
(WKViewCreateWPE):
* Source/WebKit/UIProcess/API/C/playstation/WKView.h:
* Source/WebKit/UIProcess/playstation/PageClientImpl.cpp:
(WebKit::PageClientImpl::hostFileDescriptor):
* Source/WebKit/UIProcess/playstation/PlayStationWebView.cpp:
(WebKit::PlayStationWebView::create):
(WebKit::PlayStationWebView::PlayStationWebView):
(WebKit::m_backend):
* Source/WebKit/UIProcess/playstation/PlayStationWebView.h:
(WebKit::PlayStationWebView::backend):
* Tools/MiniBrowser/playstation/CMakeLists.txt:
* Tools/MiniBrowser/playstation/WebViewWindow.cpp:
(WebViewWindow::WebViewWindow):
* Tools/MiniBrowser/playstation/WebViewWindow.h:

Canonical link: https://commits.webkit.org/264877@main
